test/integration/update_int_test.go

```diff
@@ -329,3 +347,42 @@ func (suite *UpdateIntegrationTestSuite) TestUpdateToCurrent() {
 
 	suite.testUpdate(ts, installDir, e2e.AppendEnv(fmt.Sprintf("ACTIVESTATE_CLI_UPDATE_BRANCH=%s", constants.BranchName)))
 }
+
+// repeatCheck provides methods to check a callback multiple times. This is
+// useful when a callback may not succeed on the first try, but subsequent
+// tries are expected to pass.
+type repeatCheck struct {
+	t     *testing.T
+	iters int
+	pause time.Duration
+}
+
+func newRepeatCheck(t *testing.T, iters int, pause time.Duration) *repeatCheck {
+	return &repeatCheck{
+		t:     t,
+		iters: iters,
+		pause: pause,
+	}
+}
+
+func (c *repeatCheck) contains(text func() string, substrs ...string) {
+	printf := c.t.Logf
+	pause := c.pause
+
+	for i := 0; i < c.iters; i++ {
+		if i == c.iters-1 {
+			printf = c.t.Fatalf
+			pause = 0
+		}
+
+		txt := text()
+		for _, substr := range substrs {
+			if strings.Contains(txt, substr) {
+				return
+			}
+			printf("Expected to find (iter %d):\n%q\nReceived:\n%q", i, substr, txt)
+		}
+
+		time.Sleep(pause)
+	}
+}
```

Comment:
This feels over-engineered for what is needed. It could just be a simple for loop with a break.

```
found := false
for x :=0; x<5; x++ {
		cp := ts.SpawnCmdWithOpts(stateExePath, spawnOpts...)
		cp.Wait(time.Second * 10)
		if strings.Contains(cp.Snapshot(), "Updating State Tool") {
			found = true
			break
		}
}
suite.Require().True(found, "Expecting to find 'Updating State Tool'")
``

That is roughly about as much code as you currently have to invoke the logic introduced here.

Code like this (repeatCheck) creates two problems, which is a price that may in some situations be worth paying:

- We need to maintain it
- Reading test code I now have to jump to this implementation to understand what the test is doing, where a simple for loop would've been self-explanatory